src/components/views/elements/MemberEventListSummary.js

```diff
@@ -40,100 +40,183 @@ module.exports = React.createClass({
+        //  - or if there are fewEvents, meaning the child eventTiles are shown as-is
+        return (
+            nextProps.events.length !== this.props.events.length ||
+            this.state.expanded || nextState.expanded ||
+            nextProps.events.length < this.props.threshold
+        );
+    },
+
     _toggleSummary: function() {
         this.setState({
             expanded: !this.state.expanded,
         });
     },
 
-    _getEventSenderName: function(ev) {
-        if (!ev) {
-            return 'undefined';
+    _renderNameList: function(users) {
+        if (users.length === 0) {
+            return null;
         }
-        return ev.sender.name || ev.event.content.displayname || ev.getSender();
+
+        return this._renderCommaSeparatedList(users, this.props.summaryLength);
     },
 
-    _renderNameList: function(events) {
-        if (events.length === 0) {
-            return null;
-        }
-        let originalNumber = events.length;
-        events = events.slice(0, this.props.summaryLength);
-        let lastEvent = events.pop();
-
-        let names = events.map((ev) => {
-            return this._getEventSenderName(ev);
-        }).join(', ');
-
-        let lastName = this._getEventSenderName(lastEvent);
-        if (names.length === 0) {
-            // special-case for a single event
-            return lastName;
+    // Test whether the first n items repeat for the duration
+    // e.g. [1,2,3,4,1,2,3] would resolve true for n = 4
+    _isRepeatedSequence: function(transitions, n) {
+        let count = 0;
+        for (let i = 0; i < transitions.length; i++) {
+            if (transitions[i % n] !== transitions[i]) {
+                return null;
+            }
         }
+        return true;
+    },
 
-        let remaining = originalNumber - this.props.summaryLength;
-        if (remaining > 0) {
-            //  name1, name2, name3, and 100 others
-            return names + ', ' + lastName + ', and ' + remaining + ' others';
+    _renderCommaSeparatedList(items, itemLimit) {
+        const remaining = itemLimit === undefined ? 0 : Math.max(items.length - itemLimit, 0);
+        if (items.length === 0) {
+            return "";
+        } else if (items.length === 1) {
+            return items[0];
+        } else if (remaining) {
+            items = items.slice(0, itemLimit);
+            const other = " other" + (remaining > 1 ? "s" : "");
+            return items.join(', ') + ' and ' + remaining + other;
         } else {
-            //  name1, name2 and name3
-            return names + ' and ' + lastName;
+            let last = items.pop();
+            return items.join(', ') + ' and ' + last;
         }
     },
 
-    _renderSummary: function(joinEvents, leaveEvents) {
-        let joiners = this._renderNameList(joinEvents);
-        let leavers = this._renderNameList(leaveEvents);
+    _getDescriptionForTransition(t, plural, repeats) {
```

Comment:
could do with some more comments explaining what these helpers do, and what their params are. For example, the fact that `plural` is a boolean flag meaning that more than one user was affected would be very useful knowledge here.